lib/utilities/sign.js

```diff
@@ -20,29 +32,114 @@ async function hardwareSignatures(hwApp, ptx, account, enc) {
   assert(inputCount === scripts.length, 'must have one script for each input');
   assert(inputCount === paths.length, 'must have one path for each input');
 
-  const mtx = MTX.fromJSON(tx);
-  const ledgerInputs = [];
-
-  for (const [i, input] of mtx.inputs.entries()) {
-    const { branch, index } = paths[i];
-    const path = `${account}/${branch}/${index}`;
-    const redeem = Buffer.from(scripts[i]);
-    const coin = mtx.view.getCoinFor(input);
-    const ledgerInput = new LedgerTXInput({
-      witness: true,
-      redeem,
-      coin,
-      path,
-    });
-    ledgerInputs.push(ledgerInput);
+  const mtx = tx;
+  const inputs = [];
+  const outputs = [];
+
+  switch (hwApp.type) {
+    case 'ledger':
+      {
+        for (const [i, input] of mtx.inputs.entries()) {
+          const { branch, index } = paths[i];
+          const path = `${account}/${branch}/${index}`;
+          const redeem = Buffer.from(scripts[i]);
+          const coin = mtx.view.getCoinFor(input);
+          const ledgerInput = new LedgerTXInput({
+            witness: true,
+            redeem,
+            coin,
+            path,
+          });
+          inputs.push(ledgerInput);
+        }
+      }
+      break;
+
+    case 'trezor':
+      {
+        let segwit;
+        for (const [i, input] of mtx.inputs.entries()) {
+          const { branch, index } = paths[i];
+          const path = `${account}/${branch}/${index}`;
+          // HACK
+          let amount;
+
+          const inputCoin = mtx.view.getCoin(input.prevout);
+          if (inputCoin) amount = inputCoin.value.toString();
+
+          let ti = {
+            address_n: parsePath(path, true),
+            prev_index: input.prevout.index,
+            prev_hash: input.prevout.txid(),
+            script_type: 'SPENDADDRESS',
+            amount,
+          };
+
+          segwit = false;
```

Comment:
I think there is a bug in here regarding a tx with both segwit and non segwit inputs. We need `segwit` value to determine the output address `toBase58` or `toBech32`. To have `segwit` be in scope for the `outputs` loop, its declared above and out of the loops. Its set to `false` each iteration and then set to `true` if the input `isProgram`. If the last input is legacy, then `segwit` will be `false` for the outputs loop. The last input will determine the value for the  output loop